### PR TITLE
light: CHT and bloom trie indexers working in light mode

### PR DESCRIPTION
**Pull request by zsfelfoldi**
This PR enables the indexers to work in light client mode by downloading a part of these tries (the Merkle proofs of the last values of the last known section) in order to be able to add new values and recalculate subsequent hashes. It also adds CHT data to NodeInfo.

### CODE REVIEW

File: core/chain_indexer.go [hunk 1]
```diff
@@ -44,6 +44,10 @@ type ChainIndexerBackend interface {
 
 	// Commit finalizes the section metadata and stores it into the database.
 	Commit() error
+
+	// If a functions blocks for a longer period, it should immediately return
+	// with an error when Cancel is called.
+	Cancel()
```
Comment (karalabe) [inline core/chain_indexer.go]: Does this actually terminate the indexer backend? I'm a bit worried about the semantics here, because `Cancel` is meaningful in the scope of a single request, but what happens with a background process? I could understand a `Close` that terminates everything, but the notion of Canceling some internal op seems strange as the internal state will end up corrupted.

TL;DR: Don't we really want `Close() error` here?

Comment (zsfelfoldi) [reply core/chain_indexer.go]: Shutdown is still handled by the indexer, this is just a signal for the backend to not block on external events anymore (like ODR retrievals). I agree about semantics though, I changed the name and the description to better reflect this behavior.

**Comment by zsfelfoldi** [comment]
@karalabe I changed the backend shutdown to Close(), PTAL

**Comment by dshulyak** [comment]
in case it will be helpful: i wanted to use this PR for pruning headers from older sections
but when i tried to sync headers cht indexer entered corrupted state which leads to such
this log being repeated all the time:

```
INFO [08-15|10:36:33] Upgrading chain index                    type=cht percentage=0
ERROR[08-15|10:36:17] Section processing failed                type=cht error="canonical block #1 unknown"
ERROR[08-15|10:36:17] Section processing failed                type=cht error="canonical block #1 unknown"
ERROR[08-15|10:36:17] Section processing failed                type=cht error="canonical block #1 unknown"
ERROR[08-15|10:36:17] Section processing failed                type=cht error="canonical block #1 unknown"
```

i decided that cht indexer state is corrupted, because i have another log that shows what is the last section number and section head
when SyncCht is called for synchronising with a peer:

```
INFO [08-15|10:36:15] SYNC CHT                                 head num=5848127 cht count=0   sec head=0x0000000000000000000000000000000000000000000000000000000000000000
```

Without this PR it will print adequate information such as:

```
INFO [08-15|10:44:02] SYNC CHT                                 head num=0 cht count=175 sec head=0xa3ef48cd8f1c3a08419f0237fc7763491fe89497b3144b17adf87c1c43664613
```

No special steps are required. Geth binary was build from this PR

```
./geth --light --v5disc --verbosity 3 --datadir /tmp/cht2
```

diff for a log line just in case:

```
diff --git a/light/lightchain.go b/light/lightchain.go
index a30bfb5ae..f8c48217f 100644
--- a/light/lightchain.go
+++ b/light/lightchain.go
@@ -456,7 +456,8 @@ func (self *LightChain) SyncCht(ctx context.Context) bool {
                return false
        }
        headNum := self.CurrentHeader().Number.Uint64()
-       chtCount, _, _ := self.odr.ChtIndexer().Sections()
+       chtCount, _, secHead := self.odr.ChtIndexer().Sections()
+       log.Info("SYNC CHT", "head num", headNum, "cht count", chtCount, "sec head", secHead.String())
        if headNum+1 < chtCount*CHTFrequencyClient {
                num := chtCount*CHTFrequencyClient - 1
                header, err := GetHeaderByNumber(ctx, self.odr, num)
```

**Comment by dshulyak** [comment]
so what happens is that ChainIndexer calls backend.Reset with previous section:
https://github.com/ethereum/go-ethereum/blob/master/core/chain_indexer.go#L347
in the next step it tries to initialize a trie with a root that doesn't exist in database
https://github.com/ethereum/go-ethereum/blob/master/light/postprocess.go#L155
it fails and goes into corrupted state:
https://github.com/ethereum/go-ethereum/blob/master/core/chain_indexer.go#L348

File: light/postprocess.go [hunk 2]
```diff
@@ -137,28 +141,95 @@ func NewChtIndexer(db ethdb.Database, clientMode bool) *core.ChainIndexer {
...
+			}
+		} else {
+			break
+		}
+	}
+	if err == nil {
+		r.Proof.Store(batch)
+		err = batch.Write()
+	}
+	if err == ctx.Err() {
```
Comment (dshulyak) [inline light/postprocess.go]: in case if ctx.Err() and err is nil it will still return an error

**Comment by fjl** [comment]
I'm merging this even though DB size increases because @zsfelfoldi promised to add database garbage collection soon.
